src/control/server/mgmt_svc.go

```diff
@@ -772,74 +777,41 @@ func (svc *mgmtSvc) PrepShutdownRanks(ctx context.Context, req *mgmtpb.RanksReq)
 // StopRanks implements the method defined for the Management Service.
 //
 // Stop data-plane instance managed by control-plane identified by unique rank.
-//
-// Iterate over instances issuing kill rank requests, wait until either all instances are
-// stopped or timeout occurs and then populate response results based on instance started state.
-// Return error if any instances are still running in order to enable retries at the sender.
-//
-// TODO: Enable "force" if number of retries fail, issuing a different signal/mechanism for
-//       terminating the process.
-func (svc *mgmtSvc) StopRanks(ctx context.Context, req *mgmtpb.RanksReq) (*mgmtpb.RanksResp, error) {
+// After attempting to stop instances through harness (when either all instances
+// are stopped or timeout has occurred, populate response results based on
+// instance started state.
+func (svc *mgmtSvc) StopRanks(parent context.Context, req *mgmtpb.RanksReq) (*mgmtpb.RanksResp, error) {
 	if req == nil {
 		return nil, errors.New("nil request")
 	}
 	timeout := time.Duration(req.Timeout)
+	ctx, cancel := context.WithTimeout(parent, timeout)
+	defer cancel()
 	svc.log.Debugf("MgmtSvc.StopRanks dispatch, req:%+v, timeout:%s\n", *req, timeout)
 
 	resp := &mgmtpb.RanksResp{}
 
-	for _, i := range svc.harness.instances {
-		if !i.hasSuperblock() { // critical error
-			return nil, errors.Errorf("instance %d has no superblock", i.Index())
-		}
+	signal := syscall.SIGINT
+	if req.Force {
+		signal = syscall.SIGKILL
+	}
+	if err := svc.harness.SignalInstances(ctx, svc.log, signal, req.Ranks...); err != nil {
+		return nil, errors.Wrap(err, "signalling instances to stop")
+	}
 
-		rank, ok := validateInstanceRank(svc.log, i, req.Ranks)
-		if !ok { // filtered out, no result expected
-			continue
+	// either all instances stopped or timeout occurred
+	for _, i := range svc.harness.Instances() {
+		rank, err := validateInstanceRank(svc.log, i, req.Ranks)
+		if err != nil {
+			return nil, err
```

Comment:
This seems racy. Sending the signal to all instances to shut down could succeed, but for whatever reason it may take the instances a little while to fully stop. In that case, the i.IsStarted() check would be true and then the admin would see an error that the instance(s) didn't stop. It would be more reliable to loop with a timeout, just in case the first few checks were too quick.

I also think it's a mistake to spread the logic for shutting instances down across different areas of responsibility. It means that the only way to stop the instances and verify that they're stopped is to use this method or duplicate the polling logic somewhere else. It'd be better to refactor this a bit so that the logic for synchronously shutting down is all in the harness, IMO. I'm thinking that maybe this would look like a harness.StopInstances() method which accepts a context and force boolean and encapsulates the logic for signaling/polling in one place.